pangeo_forge/recipe.py

```diff
@@ -139,6 +135,45 @@ def _chunk_metadata_fname(chunk_key) -> str:
     return "chunk-meta-" + _encode_key(chunk_key) + ".json"
 
 
+def _copy_btw_filesystems(input_opener, output_opener, BLOCK_SIZE=10_000_000):
+    with input_opener as source:
+        with output_opener as target:
+            while True:
+                data = source.read(BLOCK_SIZE)
+                if not data:
+                    break
+                target.write(data)
+
+
+@contextmanager
+def _maybe_open_or_copy_to_local(opener, copy_to_local, orig_name):
+    _, suffix = os.path.splitext(orig_name)
+    if copy_to_local:
+        ntf = tempfile.NamedTemporaryFile(suffix=suffix)
+        tmp_name = ntf.name
+        logger.info(f"Copying {orig_name} to local file {tmp_name}")
+        target_opener = open(tmp_name, mode="wb")
+        _copy_btw_filesystems(opener, target_opener)
+        yield tmp_name
+        ntf.close()  # cleans up the temporary file
+    else:
+        with opener as fp:
+            with fp as fp2:
+                yield fp2
+
+
+@contextmanager
+def _fsspec_safe_open(fname, **kwargs):
+    # workaround for inconsistent behavior of fsspec.open
+    # https://github.com/intake/filesystem_spec/issues/579
+    with fsspec.open(fname, **kwargs) as fp:
+        if isinstance(fp, fsspec.implementations.local.LocalFileOpener):
+            with fp as fp2:
+                yield fp2
+        else:
+            yield fp
```

Comment:
The need for this helper function is due to https://github.com/intake/filesystem_spec/issues/579.

Reply:
What error did you get without this? The `LocalFileOpener` supports all of the methods that an `_io.BufferedReader` would

Reply:
```
---------------------------------------------------------------------------
ValueError                                Traceback (most recent call last)
<ipython-input-77-951e7a504e4c> in <module>
      1 with of as fp:
----> 2     xr.open_dataset(fp)

~/Code/xarray/xarray/backends/api.py in open_dataset(filename_or_obj, engine, chunks, cache, decode_cf, mask_and_scale, decode_times, decode_timedelta, use_cftime, concat_characters, decode_coords, drop_variables, backend_kwargs, *args, **kwargs)
    491 
    492     if engine is None:
--> 493         engine = plugins.guess_engine(filename_or_obj)
    494 
    495     backend = plugins.get_backend(engine)

~/Code/xarray/xarray/backends/plugins.py in guess_engine(store_spec)
    108             logging.exception(f"{engine!r} fails while guessing")
    109 
--> 110     raise ValueError("cannot guess the engine, try passing one explicitly")
    111 
    112 

ValueError: cannot guess the engine, try passing one explicitly
```

Reply:
I would say this is an xarray problem: you should not require a file-like object to be a subclass of io.IOBase. In other locations, it just does `hasattr(obj, "read")` instead.

Reply:
Ok good to know. Could you raise an xarray issue for this?

Reply:
But just to clarify the source of my confusion, if I call `fsspe.open` the file with `mode='r'` (rather than `mode='rb'`), the object I get _is_ a subclass of ` io.IOBase`!

Reply:
Yes, this is true. If you wanted LocalFileOpener to be a subclass, you would need a load of methods implemented like
```python
    def read(self, size=-1):
        return self.f.read(size)
```
**or** lose the serialisable trait by making a `BufferedReader` directly.